### PR TITLE
docs: first pass at documenting how to add extensions

### PR DESCRIPTION
**Pull request by alyssawilk**
This includes enhancing the script failures to leave breadcrumbs for devs who miss the new docs.

Risk Level: n/a (tooling / docs)
Testing: manual testing removing files from #11327
Docs Changes: yes
Release Notes: no

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -215,6 +215,25 @@ and false.
 * If a PR includes a deprecation/breaking change, notification should be sent to the
   [envoy-announce](https://groups.google.com/forum/#!forum/envoy-announce) email list.
 
+# Adding new extensions
+
+For developers adding a new extension, one can largely model off of existing extensions.
```
Comment (htuch) [inline CONTRIBUTING.md:220]:
```suggestion
For developers adding a new extension, one can take an existing extension as the starting point.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -215,6 +215,25 @@ and false.
...
+
+Extension configuration should be located in a directory structure like
+`api/envoy/extensions/area/plugin/`, for example `api/envoy/extensions/tracers/zipkin/`
+
+The code for the extension should be located under the equivalent
+`source/extensions/area/plugin`, and include an envoy_cc_extension with the
+configuration and tagged with the appropriate security posture, and an
+envoy_cc_library with the code.
+
```
Comment (htuch) [inline CONTRIBUTING.md:229]: Link to https://github.com/envoyproxy/envoy/blob/master/api/STYLE.md#adding-an-extension-configuration-to-the-api

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -215,6 +215,25 @@ and false.
...
+Other changes will likely include
+
+  * Editing source/extensions/extensions_build_config.bzl to include the new extensions
```
Comment (htuch) [inline CONTRIBUTING.md:232]: Suggestion: use `:repo:` to turn these into real links.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -215,6 +215,25 @@ and false.
...
+
+For developers adding a new extension, one can largely model off of existing extensions.
+
+Extension configuration should be located in a directory structure like
+`api/envoy/extensions/area/plugin/`, for example `api/envoy/extensions/tracers/zipkin/`
+
+The code for the extension should be located under the equivalent
+`source/extensions/area/plugin`, and include an envoy_cc_extension with the
+configuration and tagged with the appropriate security posture, and an
+envoy_cc_library with the code.
```
Comment (htuch) [inline CONTRIBUTING.md:228]: Suggestion: use markup to make Bazel macros distinct, e.g. `*envoy_cc_library*`.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -215,6 +215,26 @@ and false.
 * If a PR includes a deprecation/breaking change, notification should be sent to the
   [envoy-announce](https://groups.google.com/forum/#!forum/envoy-announce) email list.
 
+# Adding new extensions
+
+For developers adding a new extension, one can take an existing extension as the starting point.
+
+Extension configuration should be located in a directory structure like
+`api/envoy/extensions/area/plugin/`, for example `api/envoy/extensions/tracers/zipkin/`
```
Comment (lizan) [inline CONTRIBUTING.md:223]: Let's use other extension as examples other than tracers, they only have `v4alpha` which might be confusing.

**Comment by alyssawilk** [comment]
@htuch I miss anything or is this good to go?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -215,6 +215,26 @@ and false.
...
+For developers adding a new extension, one can take an existing extension as the starting point.
+
+Extension configuration should be located in a directory structure like
+`api/envoy/extensions/area/plugin/`, for example `api/envoy/extensions/access_loggers/file/`
+
+The code for the extension should be located under the equivalent
+`source/extensions/area/plugin`, and include an *envoy_cc_extension* with the
+configuration and tagged with the appropriate security posture, and an
+*envoy_cc_library* with the code. More details on how to add a new extension
+API can be found [here](api/STYLE.md):
```
Comment (htuch) [inline CONTRIBUTING.md:229]:
```suggestion
API can be found [here](api/STYLE.md#adding-an-extension-configuration-to-the-api):
```

**Comment by alyssawilk** [comment]
Nit should be addressed!

**Pull request merged**